pystassh/session.py

```diff
@@ -57,51 +61,59 @@ def connect(self):
-                        ret, self.get_error_message(session)))
-        else:
-            ret = api.Api.ssh_userauth_autopubkey(session, self._passphrase)
-            if ret != api.SSH_AUTH_SUCCESS:
-                raise exceptions.AuthenticationException(
-                    "Authentication cannot be made with public key (return code: {}): {}".format(
-                        ret, self.get_error_message(session)))
-
-        self._session = session
-        self._channel = Channel(self._session)
+        try:
+            ret = self._api.ssh_options_set(session, api.SSH_OPTIONS_HOST, self._hostname)
+            if ret != api.SSH_OK:
+                raise exceptions.ConnectionException("Hostname '{}' cannot be set (return code: {}): {}".format(
+                    self._hostname, ret, self.get_error_message(session)))
+
+            ret = self._api.ssh_options_set(session, api.SSH_OPTIONS_PORT_STR, self._port)
+            if ret != api.SSH_OK:
+                raise exceptions.ConnectionException("Port '{}' cannot be set (return code: {}): {}".format(
+                    self._port, ret, self.get_error_message(session)))
+
+            ret = self._api.ssh_options_set(session, api.SSH_OPTIONS_USER, self._username)
+            if ret != api.SSH_OK:
+                raise exceptions.ConnectionException("Username '{}' cannot be set (return code: {}): {}".format(
+                    self._username, ret, self.get_error_message(session)))
+
+            ret = self._api.ssh_connect(session)
+            if ret != api.SSH_OK:
+                raise exceptions.ConnectionException("Connection cannot be made (return code: {}): {}".format(
+                    ret, self.get_error_message(session)))
+
+            if self._password:
+                ret = self._api.ssh_userauth_password(session, self._username, self._password)
+                if ret != api.SSH_AUTH_SUCCESS:
+                    raise exceptions.AuthenticationException(
+                        "Authentication cannot be made with username and password (return code: {}): {}".format(
+                            ret, self.get_error_message(session)))
+            else:
+                ret = self._api.ssh_userauth_autopubkey(session, self._passphrase)
+                if ret != api.SSH_AUTH_SUCCESS:
+                    raise exceptions.AuthenticationException(
+                        "Authentication cannot be made with public key (return code: {}): {}".format(
+                            ret, self.get_error_message(session)))
+
+            self._session = session
+            self._channel = Channel(self._session)
+        except:
+            self._api.ssh_free(session)
+            self._session = self._channel = None
+            raise
 
     def disconnect(self):
         """ Close the current connection.
         """
         if self.is_connected():
-            api.Api.ssh_disconnect(self._session)
+            self._channel.close()
```

Comment:
I got a warning with this line while running the tests, because `self._channel` was `None`. I changed it to only call `close` when it's not.

Reply:
Nice!